File: algorithms/python/specialBinaryString/specialBinaryString.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/special-binary-string/
# Author: Miao Zhang
# Date:   2021-03-08

class Solution:
    def makeLargestSpecial(self, S: str) -> str:
        cnt = 0
        i = 0
        v = []
        for j in range(len(S)):
            cnt += 1 if S[j] == '1' else -1
            if cnt == 0:
                v.append('1' + self.makeLargestSpecial(S[i + 1: j]) + '0')
                i = j + 1
        v.sort(reverse=True)
        return ''.join(v)

File: algorithms/cpp/specialBinaryString/specialBinaryString.cpp
// Source: https://leetcode.com/problems/special-binary-string/
// Author: Miao Zhang
// Date:   2021-03-08

class Solution {
public:
    string makeLargestSpecial(string S) {
        int cnt = 0, i = 0;
        vector<string> v;
        for (int j = 0; j < S.size(); j++) {
            cnt += (S[j] == '1') ? 1 : -1;
            if (cnt == 0) {
                v.push_back('1' + makeLargestSpecial(S.substr(i + 1, j - i - 1)) + '0');
                i = j + 1;
            }
        }
        sort(v.begin(), v.end(), greater<string>());
        string res = "";
        for (int i = 0; i < v.size(); i++)  res += v[i];
        return res;
    }
};
